### PR TITLE
[Console] Support binary / negatable options

### LINKED ISSUE #24314: [Console] Support for --foo / --no-foo binary options
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 3.x

## Problem

We are using `$inputOption->setDefault($value);` to inject [configuration settings into Symfony Console applications](https://github.com/consolidation/config). The advantage of doing this is that the current configuration value is shown in the help output as the default value, and specifying a different value via an option on the commandline overrides the configuration value, as it should do.

However, if we call `setDefault()` on a boolean option, we get an error message.

## Proposed solution

Some (non-Symfony-Console) commandline tools display option help like this:
```
    -i, --[no-]interactive           be interactive and colorful
```
The above example is from the [Travis CLI tool](https://github.com/travis-ci/travis.rb).

A default value could be added to these for consistency:
```
    -i, --[no-]interactive           be interactive and colorful [default: --interactive]
```

## Considerations for backwards compatibility

In order to preserve backwards compatibility, a new InputOption constant could be added to the existing set (`InputOption::VALUE_NONE`, `InputOption::VALUE_REQUIRED`, etc.) to enable the new behavior.  Adding `InputOption::VALUE_BINARY` would be a drop-in replacement for `InputOption::VALUE_NONE` that would have the following effect:

- The help text for the option would be rendered as shown above.
- Setting a default value would now be allowed.
- The default value for the option would have to be boolean, and would `throw` if not.
- Input validation would automatically support a `--no-foo` option for any defined binary `--foo`. The result of supplying `--no-foo` would be to set the value of the `--foo` option to `false`.

The behavior of supplying `--foo` and `--no-foo` at the same time should be deterministic, but there are options on how this is handled. I might favor "last one wins", but throwing an exception would also be possible.

## Requested action

We would like to know if a pull request related to this feature would be considered, and if so, would it be eligible for eventual release on the Symfony 3.x branch?

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #24314
| License       | MIT
| Doc PR        | TODO

This PR take over #26292 to introduce a new option mode `NEGATABLE` that ease creation of option and their negation (ie. `--foo`, `--no-foo`)

Negated options applies only to flag options (`VALUE_NONE`)
```
./cli.php --foo
var_dump(getOption('foo')); // true

./cli.php --no-foo
var_dump(getOption('foo')); // false
```

### CODE REVIEW

**Comment by fabpot** [comment]
I would only allow the `--no-` prefix for options that act as flags (the ones that do not take a value).

File: src/Symfony/Component/Console/Input/InputDefinition.php [hunk 1]
```diff
@@ -297,6 +307,14 @@ public function hasShortcut(string $name)
         return isset($this->shortcuts[$name]);
     }
 
+    /**
+     * Returns true if an InputOption object exists by negated name.
```
Comment (OskarStark) [inline src/Symfony/Component/Console/Input/InputDefinition.php:311]:
```suggestion
     * Returns true if an InputOption object exists and is negatable.
```
?

File: src/Symfony/Component/Console/Input/InputDefinition.php [hunk 2]
```diff
@@ -338,6 +356,22 @@ public function shortcutToName(string $shortcut): string
         return $this->shortcuts[$shortcut];
     }
 
+    /**
+     * Returns the InputOption name given a negation.
+     *
+     * @throws InvalidArgumentException When option given does not exist
```
Comment (OskarStark) [inline src/Symfony/Component/Console/Input/InputDefinition.php:362]:
```suggestion
     * @throws InvalidArgumentException When given option does not exist
```

File: src/Symfony/Component/Console/Tests/Command/ListCommandTest.php [hunk 3]
```diff
@@ -80,8 +80,7 @@ public function testExecuteListsCommandsOrder()
   -h, --help            Display help for the given command. When no command is given display help for the list command
   -q, --quiet           Do not output any message
   -V, --version         Display this application version
-      --ansi            Force ANSI output
-      --no-ansi         Disable ANSI output
+      --[no-]ansi       Force ANSI output
```
Comment (OskarStark) [inline src/Symfony/Component/Console/Tests/Command/ListCommandTest.php]: Before the no option has "Disable ANSI...." now you can understand it as "do not force it, but do not disable it"

Shall we add add some magic for the description? Or shall we add sth to provide a description for the negatable option?

File: src/Symfony/Component/Console/Input/InputDefinition.php (hunk 1 above)
Comment (jderusse) [reply src/Symfony/Component/Console/Input/InputDefinition.php:311]: hmm.. if you send a suggestion, it means that the original comment is missleading. But your suggestion is wrong. Please help me finding a better description.

Here is the situation: we have options that are negatable. The negated name of `--foo` is `--no-foo`.

This function, returns true when called with argument `--no-foo`. This comment is very similar to `hasShortcut` method `Returns true if an InputOption object exists by shortcut.`

File: src/Symfony/Component/Console/Tests/Command/ListCommandTest.php (hunk 3 above)
Comment (jderusse) [reply src/Symfony/Component/Console/Tests/Command/ListCommandTest.php]: You're right. What's about?
* `Force/Disable ANSI output`
* `Force [or Disable] ANSI output`

File: src/Symfony/Component/Console/Input/InputDefinition.php (hunk 2 above)
Comment (jderusse) [reply src/Symfony/Component/Console/Input/InputDefinition.php:362]: Fix other occurrences for consistency

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.xml [hunk 4]
```diff
@@ -24,34 +24,31 @@
...
-        <option name="--version" shortcut="-V" accept_value="0" is_value_required="0" is_multiple="0">
+        <option name="--version" shortcut="-V" accept_value="0" is_value_required="0" is_multiple="0" is_negatable="0">
           <description>Display this application version</description>
         </option>
-        <option name="--ansi" shortcut="" accept_value="0" is_value_required="0" is_multiple="0">
-          <description>Force ANSI output</description>
+        <option name="--ansi" shortcut="" accept_value="0" is_value_required="0" is_multiple="0" is_negatable="1">
+          <description>Force/Disable ANSI output</description>
         </option>
-        <option name="--no-ansi" shortcut="" accept_value="0" is_value_required="0" is_multiple="0">
```
Comment (jderusse) [inline src/Symfony/Component/Console/Tests/Fixtures/application_1.xml:51]: Is it a BC break? Scripts that use the xml (or json) descriptor to provide auto-completion won't know that `--no-XXX` exists until they supports the `is_negatable` attribute.

File: src/Symfony/Component/Console/Input/InputDefinition.php (hunk 1 above)
Comment (OskarStark) [reply src/Symfony/Component/Console/Input/InputDefinition.php:311]: Sorry you are right, I Mixed things up.

File: src/Symfony/Component/Console/Tests/Command/ListCommandTest.php (hunk 3 above)
Comment (OskarStark) [reply src/Symfony/Component/Console/Tests/Command/ListCommandTest.php]: I like the second one 👍🏻

**Review by OskarStark — APPROVED** [review]

**Comment by OskarStark** [comment]
Can you please make fabbot happy? 😊

**Comment by jderusse** [comment]
> Can you please make fabbot happy? 😊

I think, fabbot is a false positive here. 
Fixing it will break the test

**Comment by OskarStark** [comment]
Indeed it looks weird 🧐

**Comment by javiereguiluz** [comment]
Jérémy thanks for this contribution.

As a super minor comment: is "negatable" the best word to describe this? Could we use "boolean" like the Golang boolean flags?

File: src/Symfony/Component/Console/Tests/Command/ListCommandTest.php [hunk 5]
```diff
@@ -80,8 +80,7 @@ public function testExecuteListsCommandsOrder()
   -h, --help            Display help for the given command. When no command is given display help for the list command
   -q, --quiet           Do not output any message
   -V, --version         Display this application version
-      --ansi            Force ANSI output
-      --no-ansi         Disable ANSI output
+      --[no-]ansi       Force [or Disable] ANSI output
```
Comment (wouterj) [inline src/Symfony/Component/Console/Tests/Command/ListCommandTest.php]: I think we should make this `--ansi, --no-ansi  Force [or disable] ANSI output`. The text describer in Symfony is [docopt](http://docopt.org/) compatible and `--[no-]ansi` is not part of the docopt spec.

Or are there other common conventions (e.g. Linux) that apply this type of syntax?

**Comment by OskarStark** [comment]
I like the Boolean naming a lot! Thank you for the proposal Javier 😃

**Comment by chalasr** [comment]
Boolean looks confusing IMO.
Actually, Go' boolean flags are not about options that have a `--no-` alternative, but just options whose value can either be `true` or `false`.
Negatable options can be found in [Perl](https://perldoc.perl.org/Getopt::Long#A-little-bit-less-simple-options) and [node](https://github.com/tj/commander.js/) and other languages. I would stick with it.

**Comment by jderusse** [comment]
~renamed VALUE_NEGATABLE into VALUE_BOOLEAN~

I also Replaced `--[no-]foo` notation by `--foo|no-foo` for consistency with shortcut `-v|vv|vvv` and respects docopt  (`|` means either X or X).
But it lloks like we are not 100% compliant with the spec: 
* `-v|vv|vvv` should be `-v | -vv | -vvv`, 
* `-e, --env=ENV` should be `-e --env=ENV` or `-e, --env <env>`

**Comment by wouterj** [comment]
You're correct about `-v|vv|vvv` (let's fix that in another PR), but `-e, --env=ENV` is correct afaics: _"Follow either `<angular-brackets>` or `UPPER-CASE` convention for options' arguments. You can use a comma if you want to separate options."_ Seems like all variations of comma/no-comma, and `<angular-brackets>`/`UPPER-CASE` are valid.

(and sorry for being nitpicky about this, I just remember this as it was my first big PR to Symfony: #13220)

File: src/Symfony/Component/Console/Input/InputOption.php [hunk 6]
```diff
@@ -82,6 +83,9 @@ public function __construct(string $name, $shortcut = null, int $mode = null, st
         if ($this->isArray() && !$this->acceptValue()) {
             throw new InvalidArgumentException('Impossible to have an option mode VALUE_IS_ARRAY if the option does not accept a value.');
         }
+        if ($this->isNegatable() && $this->acceptValue()) {
+            throw new InvalidArgumentException('Impossible to have an option mode VALUE_NEGATABLE if the option also accept a value.');
```
Comment (chalasr) [inline src/Symfony/Component/Console/Input/InputOption.php]:
```suggestion
            throw new InvalidArgumentException('Impossible to have an option mode VALUE_NEGATABLE if the option also accepts a value.');
```

File: src/Symfony/Component/Console/Application.php [hunk 7]
```diff
@@ -1033,8 +1033,7 @@ protected function getDefaultInputDefinition()
             new InputOption('--quiet', '-q', InputOption::VALUE_NONE, 'Do not output any message'),
             new InputOption('--verbose', '-v|vv|vvv', InputOption::VALUE_NONE, 'Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug'),
             new InputOption('--version', '-V', InputOption::VALUE_NONE, 'Display this application version'),
-            new InputOption('--ansi', '', InputOption::VALUE_NONE, 'Force ANSI output'),
-            new InputOption('--no-ansi', '', InputOption::VALUE_NONE, 'Disable ANSI output'),
+            new InputOption('--ansi', '', InputOption::VALUE_NEGATABLE, 'Force [or Disable] ANSI output', null),
```
Comment (fabpot) [inline src/Symfony/Component/Console/Application.php]: What about `Force (or disable --no-ansi) ANSI output`.
Or maybe we should automatically add something when an option is negatable: `Force ANSI output (can be disabled via --no-ansi)` where the `(can be disabled via --no-ansi)` part is automatically added by Symfony?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Console/Application.php (hunk 7 above)
Comment (OskarStark) [reply src/Symfony/Component/Console/Application.php]: I would love that kind of magic here 👍🏻

Comment (jderusse) [reply src/Symfony/Component/Console/Application.php]: mhhh.. what's about people that use a non-english description ?

I wouldn't use magic here and let people write the full sentences.

edit: and maybe the word `disable` will not always fit well for the `-no` option.

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.xml (hunk 4 above)
Comment (chalasr) [reply src/Symfony/Component/Console/Tests/Fixtures/application_1.xml:51]: Works for me as-is.

File: src/Symfony/Component/Console/Application.php (hunk 7 above)
Comment (jderusse) [reply src/Symfony/Component/Console/Application.php]: I changed the sentence to `(or disable --no-ansi)`.

But I'm really not conviced by automated sentence:
- won't work for non-english application
- wording may vary with context (does disable (or whatever choosed word) will always work for this option?
- it's easy for the user to enter the right sentence.

File: src/Symfony/Component/Console/Input/ArgvInput.php [hunk 8]
```diff
@@ -204,12 +204,22 @@ private function addShortOption(string $shortcut, $value)
     /**
      * Adds a long option value.
      *
-     * @throws RuntimeException When option given doesn't exist
+     * @throws RuntimeException When given option doesn't exist
```
Comment (chalasr) [inline src/Symfony/Component/Console/Input/ArgvInput.php]: Unrelated fixes should be reverted

**Review by chalasr — COMMENTED** [review]
The changelog needs to be updated

File: src/Symfony/Component/Console/Application.php (hunk 7 above)
Comment (OskarStark) [reply src/Symfony/Component/Console/Application.php]: I agree 👍🏻

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Console/Input/InputOption.php [hunk 9]
```diff
@@ -146,6 +150,16 @@ public function isArray()
         return self::VALUE_IS_ARRAY === (self::VALUE_IS_ARRAY & $this->mode);
     }
 
+    /**
+     * Returns true if the option is negatable.
+     *
+     * @return bool true if mode is self::VALUE_NEGATABLE, false otherwise
+     */
+    public function isNegatable()
```
Comment (fabpot) [inline src/Symfony/Component/Console/Input/InputOption.php]: `: bool`?

And I would remove the whole docblock that does not seem to be needed.

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.xml (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/Console/Tests/Fixtures/application_1.xml:51]: It IS a BC break and as IDEs are using these files, I think we need to keep the current format and remove the is_negatable attribute (same for all machine consumable formats). That way, there is no need for consumers to support the new options, and having two entries for negatable options make sense to me.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.xml (hunk 4 above)
Comment (jderusse) [reply src/Symfony/Component/Console/Tests/Fixtures/application_1.xml:51]: restored. Please have a look to the `--no-XXX` description's wording

File: src/Symfony/Component/Console/Descriptor/XmlDescriptor.php [hunk 10]
```diff
@@ -225,6 +225,17 @@ private function getInputOptionDocument(InputOption $option): \DOMDocument
...
 
+        if ($option->isNegatable()) {
+            $dom->appendChild($objectXML = $dom->createElement('option'));
+            $objectXML->setAttribute('name', '--no-'.$option->getName());
+            $objectXML->setAttribute('shortcut', '');
+            $objectXML->setAttribute('accept_value', 0);
+            $objectXML->setAttribute('is_value_required', 0);
+            $objectXML->setAttribute('is_multiple', 0);
+            $objectXML->appendChild($descriptionXML = $dom->createElement('description'));
+            $descriptionXML->appendChild($dom->createTextNode('negate the "--'.$option->getName().'" option'));
```
Comment (fabpot) [inline src/Symfony/Component/Console/Descriptor/XmlDescriptor.php]: Negate (instead of negate)

**Review by fabpot — CHANGES_REQUESTED** [review]
Small nitpick before merging

File: src/Symfony/Component/Console/Descriptor/XmlDescriptor.php (hunk 10 above)
Comment (jderusse) [reply src/Symfony/Component/Console/Descriptor/XmlDescriptor.php]: done

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @greg-1-anderson.

**Comment by chalasr** [comment]
Thanks @jderusse also :)

File: src/Symfony/Component/Console/Input/InputDefinition.php [hunk 11]
```diff
@@ -246,6 +248,14 @@ public function addOption(InputOption $option)
                 $this->shortcuts[$shortcut] = $option->getName();
             }
         }
+
+        if ($option->isNegatable()) {
+            $negatedName = 'no-'.$option->getName();
+            if (isset($this->options[$negatedName])) {
+                throw new LogicException(sprintf('An option named "%s" already exists.', $negatedName));
```
Comment (stof) [inline src/Symfony/Component/Console/Input/InputDefinition.php:255]: this validation misses a case: when adding the explicit `no-foo` option *after* the negated option.

Comment (jderusse) [reply src/Symfony/Component/Console/Input/InputDefinition.php:255]: good catch. Fixed by #39723

**Comment by greg-1-anderson** [comment]
Is a backport to Symfony 4.4.x permissible?

**Comment by stof** [comment]
@greg-1-anderson no. New features are never backported in patch releases.

File: src/Symfony/Component/Console/Application.php [hunk 12]
```diff
@@ -1033,8 +1033,7 @@ protected function getDefaultInputDefinition()
             new InputOption('--quiet', '-q', InputOption::VALUE_NONE, 'Do not output any message'),
             new InputOption('--verbose', '-v|vv|vvv', InputOption::VALUE_NONE, 'Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug'),
             new InputOption('--version', '-V', InputOption::VALUE_NONE, 'Display this application version'),
-            new InputOption('--ansi', '', InputOption::VALUE_NONE, 'Force ANSI output'),
-            new InputOption('--no-ansi', '', InputOption::VALUE_NONE, 'Disable ANSI output'),
```
Comment (driesvints) [inline src/Symfony/Component/Console/Application.php:1037]: Why was this flag removed?

**Comment by driesvints** [comment]
@jderusse @fabpot the `no-ansi` flag was removed in this PR and broke the Laravel installer: https://github.com/laravel/installer/pull/202. That seems like a breaking change to me on a minor release? Although it seems like we can work around it for now easily.

**Comment by stof** [comment]
Well, was Lavaral checking the `no-ansi` option directly instead of checking whether the output was decorated ? If yes, it was not dealing with the smart default when neither `--no-ansi` or `--ansi` was passed.
The BC promise for application-level options is for the caller (and there, `--no-ansi` still works as `--ansi` is now a negatable option).

**Comment by driesvints** [comment]
We switched from `getOption` to `hasOption`:

Screen Shot 2021-06-01 at 10 42 56

**Comment by stof** [comment]
@driesvints this is not equivalent. `hasOption` tells you whether the input has a definition for this option. It won't tell you whether the option was used when calling the command.

**Comment by stof** [comment]
what you really need to check is `$output->isDecorated()` to pass either `--ansi` or `--no-ansi` to force the subcommand to have the same decoration.

**Comment by driesvints** [comment]
@stof never have had to do that before. This has always worked like this. I still feel this is a breaking change, sorry.

**Comment by driesvints** [comment]
@stof I can't seem to find any docs about `isDecorated` on symfony.com. How would one use something like that?

**Comment by stof** [comment]
@driesvints there is no documentation about accessing `--no-ansi` directly in your own command either. and as I said, passing an explicit `--no-ansi` is not the only way to get a decorated output. The default is a smart one (which is why we also have a `--ansi` to force colors).
And the case of executing a command in a subprocess like you do is precisely the case where the smart default will probably be different in the subprocess (probably never enabling colors).

> How would one use something like that?

call `$output->isDecorated()` (which returns a boolean) to know whether messages will be decorated (i.e. whether they support being formatted with color, bold, etc...)

**Comment by driesvints** [comment]
Thanks @stof, that was clear 👍
